src/main/harbour.1
.TH HARBOUR 1

.SH NAME
harbour \- Harbour compiler

.SH SYNOPSIS
\fBharbour\fP \fB<file[s][.prg]|@file>\fP \fB[options]\fP

.SH DESCRIPTION
\fBharbour\fP is a CA-Cl*pper compatible compiler for multiple platforms.

.SH OPTIONS
.IP "\fB-a\fP" 10
Automatic memvar declaration
.IP "\fB-b\fP" 10
Debug info
.IP "\fB-build\fP" 10
display detailed version info
.IP "\fB-credits\fP" 10
display credits
.IP "\fB-d<id>[=<val>]\fP" 10
#define <id>
.IP "\fB-es[<level>]\fP" 10
set exit severity
.IP "\fB-g<type>\fP" 10
output type generated is <type> (see below)
.IP "\fB-gc[<type>]\fP" 10
output type: C source (.c) (default)
.IP
<type>: 0=compact (default) 1=normal 2=verbose
.IP
3=generate real C code instead of PCODE
.IP "\fB-gh\fP" 10
output type: Harbour Portable Object (.hrb)
.IP "\fB-gd[.<destext>]\fP" 10
generate dependencies list into (.d) file
.IP "\fB-ge[<mode>]\fP" 10
error output <mode>: 0=Clipper (default) 1=IDE friendly
.IP "\fB-i<path>\fP" 10
#include file search path
.IP "\fB-i[-|+]\fP" 10
disable/enable support for INCLUDE envvar
.IP "\fB-j[<file>]\fP" 10
generate i18n gettex file (.pot)
.IP "\fB-k\fP" 10
compilation mode (type -k? for more data)
.IP "\fB-l\fP" 10
suppress line number information
.IP "\fB-m\fP" 10
compile module only
.IP "\fB-n[<type>]\fP" 10
no implicit starting procedure (default)
.IP
<type>: 0=no implicit starting procedure
.IP
1=no starting procedure at all
.IP
2=add starting procedure if necessary
.IP "\fB-o<path>\fP" 10
object file drive and/or path
.IP "\fB-p[<path>]\fP" 10
generate pre-processed output (.ppo) file
.IP "\fB-p+\fP" 10
generate pre-processor trace (.ppt) file
.IP "\fB-q\fP" 10
quiet
.IP "\fB-q0\fP" 10
quiet and don't display program header
.IP "\fB-q2\fP" 10
disable all output messages
.IP "\fB-r=<max>\fP" 10
set maximum number of preprocessor iterations
.IP "\fB-s[m]\fP" 10
syntax check only [minimal for dependencies list]
.IP "\fB-u[<file>]\fP" 10
use command def set in <file> (or none)
.IP "\fB-u+<file>\fP" 10
add command def set from <file>
.IP "\fB-undef:<id>\fP" 10
#undef <id>
.IP "\fB-v\fP" 10
variables are assumed M->
.IP "\fB-w[<level>]\fP" 10
set warning level number (0..3, default 1)
.IP "\fB-x[<prefix>]\fP" 10
set symbol init function name prefix (for .c only)
.IP "\fB-z\fP" 10
suppress shortcutting (.and. & .or.)
.IP "\fB@<file>\fP" 10
compile list of modules in <file>

.SH AUTHOR

Authors of Harbour: Type 'harbour -credits' for the Harbour Team

This manual page was written by Luis Mayoral <mayoral@linuxadicto.org>,
for the Debian GNU/Linux system (but may be used by others).


harbour/NEWS
/*
 * $Id$
 */

/* This work is licensed under the Creative Commons Attribution-ShareAlike
   License. To view a copy of this license, visit
   http://creativecommons.org/licenses/by-sa/3.0/ or send a letter to
   Creative Commons, 559 Nathan Abbott Way, Stanford, California 94305,
   USA.  A summary of the license and the full legal text is included
   in file: COPYING.  If you wish to distribute some or all of this work
   under different terms, please contact respective author(s).
   The license applies to all entries on or after Version 1.1.0.
*/

Version 3.0.0 (2011-07-17)

Note
=============================

  [TODO]

  Please refer to the ChangeLog in the meantime, particularly
  to entries marked with the keyword "INCOMPATIBILITY".

Known build issues
=============================
- None

Known open issues
=============================
- HB_TCHAR_* macro usage in Harbour sources
- ACHOICE() minor Clipper compatibility bugs
- MENUSYS minor Clipper incompatibility
- Target specific C code generated for some numeric
  operations in -gc3 mode (preventing cross-compilation)
- HBQT class creation not MT safe for multiple reasons (f.e. strtok)
- HBQT valgrind test shows double free at exit
- GTWVW bugs (scrolling bug still there)


Version 2.0.0 (2009-12-22)

Note
=============================

Too many to list, it's yet to be collected and sorted in
digestible form.

[ Incomplete, unordered.
  This section only contains update between
  2008-12-19 19:46 UTC+0100 and 2009-03-11 08:51 UTC+0100.

  2008-08/09 - 2008-12-19 and
  2009-03-11 - 2009-12-22 are missing. ]

General
=============================
- Build configuration changes:
  - PRG_USR             -> HB_USER_PRGFLAGS
  - C_USR               -> HB_USER_CFLAGS
  - L_USR               -> HB_USER_LDFLAGS
  - A_USR               -> HB_USER_AFLAGS
  - MK_USR              -> HB_USER_MAKEFLAGS
  - HB_ARCHITECTURE     -> HB_PLATFORM
  - HB_PLATFORM=w32     -> HB_PLATFORM=win
  - CCPATH              -> HB_CCPREFIX
  - CCPREFIX            -> HB_CCPATH
- GNU-make system revised to give optimal support -j<n>
  multi-threaded parallel builds, which can dramatically
  increase build speed in multi-core/multi-CPU systems.
- HB_WINCE          -> HB_OS_WIN_CE
- HB_OS_WIN_32_USED -> HB_OS_WIN_USED
- HB_OS_WIN_32      -> HB_OS_WIN
- Cleaned references of '32' in context of Windows and
  Windows API.
- New macro names introduced:
  - SUCCESS -> HB_SUCCESS
  - FAILURE -> HB_FAILURE
  - ERRCODE -> HB_ERRCODE
  (old ones still work for compatibility, but please
  switch your code to use the new ones)
- Fixed dynamic library creation on Solaris OSes.
- hbrun will now open .dbf files when passed on the command line.
- Added web-server written in Harbour: uHTTPD
- POCC WinCE builds are now supported.
- Open Watcom 1.8 (RC3) compiler is now supported.
- Removed HB_LEGACY_LEVEL support.
- #define HB_FM_WIN32_ALLOC renamed to HB_FM_WIN_ALLOC.
- Protected incompatible changes with HB_LEGACY_LEVEL2.
  (HB_SUCCESS, HB_FAILURE, HB_ERRCODE, HB_WINCE, HB_OS_WIN_32*,
  __RDDGETTEMPALIAS(), __HRB*(), some class functions,
  Harbour -go option)
  These features are still available, and will be removed
  in next major version (Harbour 1.2). To test your code for
  compatibility, you can turn it off manually by #defining
  HB_LEGACY_OFF. (HB_USER_CFLAGS=-DHB_LEGACY_OFF)
- New Harbour homepage, thanks to Vailton Renato.
  http://harbour-project.org/
- GNU Make system now supports parallel builds.
  (needs GNU Make newer than 3.79.1)
- GNU Make system allows to install in Harbour source tree on all platforms.
- GNU Make system for bcc and msvc options synced with non-GNU build methods.
- GNU Make system no longer needs -r parameter.
- GNU Make system will now generate hbfm and hbfmmt lib for all compilers.
  Added few more similar unification cleanups.
- Implemented .dll generation for msvc, msvc64, bcc and watcom when using
  GNU Make system.
- Added support for Open Watcom 1.8 (it's recommended over 1.7).
- Added support for import lib generation when using GNU Make with msvc and vcc.
  Enable it with 'set HB_DIR_IMPLIB=yes' plus set HB_DIR_* envvars.
- -mtune=pentiumpro optimization is now default for mingw/cygwin.
- Cleaned new mingw 4.3.3 warnings.
- Certain features will require Windows NT or upper when building
  Harbour on Windows, so this is now a requirement for a proper build.
- hb-mkslib.sh renamed to hb-mkdyn.sh. Old name stil works, but it's
  now deprecated.
- x64 msvc compiler got a distinct HB_COMPILER value: msvc64.
- msvcce compiler made compatible with older MSVC/WinCE versions.
- Added support for Intel(R) C++ compiler (icc) on Linux and Windows.
- Added new README document giving a quick describing of the build
  process on all supported platforms and compilers.
- Several steps to make the build process simpler and requiring the
  less amount of preparation.
- Fixed missing watcom support for HB_USER_LDFLAGS.
- Added support for MSVC IA64 mode. (untested)
- Added support for HB_CONTRIBLIBS=no to disable building of all contribs.
- Added ability to generate Windows installer and .zip distribution file.
- Fully transitioned to GNU Make system for all compilers.
  Now msvc, msvcce and bcc builds need GNU Make, see README for more
  information. "non-GNU" make systems got deleted from the source tree.
- Cygwin renamed from gcc to cygwin. (NOTE: Cygwin doesn't work since
  some versions, and support may be removed in the future.)
- Fixed #pragma -k? options to be case-insensitive.
- Explicitly bind system libraries with harbour shared library.
  RPM/DEB packages should automatically detect these dependencies.
- Warning cleanup (minors) for Intel compiler.
- Pelles C workarounds for compiler bugs.
- Added support for extended definition files in Harbour compiler
  switches: -u+<file>
- Fixed memory leak when repeated -u[<file>] options are used.

Core
=============================
- Added MT (Multi thread) support.

src/main/harbour.c
/*
 * Harbour standalone compiler entry point
 *
 * Copyright 2007 Przemyslaw Czerpak <druzus / at / priv.onet.pl>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; see the file COPYING.txt.  If not, write to
 * the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA (or visit the web site https://www.gnu.org/).
 *
 * As a special exception, the Harbour Project gives permission for
 * additional uses of the text contained in its release of Harbour.
 *
 * The exception is that, if you link the Harbour libraries with other
 * files to produce an executable, this does not by itself cause the
 * resulting executable to be covered by the GNU General Public License.
 * Your use of that executable is in no way restricted on account of
 * linking the Harbour library code into it.
 *
 * This exception does not however invalidate any other reasons why
 * the executable file might be covered by the GNU General Public License.
 *
 * This exception applies only to the code released by the Harbour
 * Project under the name Harbour.  If you copy code from other
 * Harbour Project or Free Software Foundation releases into a copy of
 * Harbour, as the General Public License permits, the exception does
 * not apply to the code that you add in this way.  To avoid misleading
 * anyone as to the status of such modified files, you must delete
 * this exception notice from them.
 *
 * If you write modifications of your own for Harbour, it is your choice
 * whether to permit this exception to apply to your modifications.
 * If you do not wish that, delete this exception notice.
 *
 */

#include "hbcomp.h"
#include "hbverbld.h"

/* Source repository ID string */
const char * hb_verCommitID( void )
{
   return HB_VER_COMMIT_ID;
}

/* Source repository revision number */
int hb_verCommitRev( void )
{
   return HB_VER_COMMIT_REV;
}

/* Last commit string */
const char * hb_verCommitInfo( void )
{
   return HB_VER_COMMIT_INFO;
}

/* build time C compiler flags in HB_USER_CFLAGS envvar */
const char * hb_verFlagsC( void )
{
#ifdef HB_VER_HB_USER_CFLAGS
   return HB_VER_HB_USER_CFLAGS;
#else
   return "";
#endif
}

/* build time linker flags in HB_USER_LDFLAGS envvar */
const char * hb_verFlagsL( void )
{
#ifdef HB_VER_HB_USER_LDFLAGS
   return HB_VER_HB_USER_LDFLAGS;
#else
   return "";
#endif
}

/* build time Harbour compiler flags in HB_USER_PRGFLAGS envvar */
const char * hb_verFlagsPRG( void )
{
#ifdef HB_VER_HB_USER_PRGFLAGS
   return HB_VER_HB_USER_PRGFLAGS;
#else
   return "";
#endif
}


#if 0
/* build time Harbour platform setting */
const char * hb_verHB_PLAT( void )
{
#ifdef HB_PLATFORM
   return HB_PLATFORM;
#else
   return "";
#endif
}

/* build time Harbour compiler setting */
const char * hb_verHB_COMP( void )
{
#ifdef HB_COMPILER
   return HB_COMPILER;
#else
   return "";
#endif
}
#endif

int main( int argc, char * argv[] )
{
   int iResult;

   hb_compChkFileSwitches( argc, argv );

   iResult = hb_compMain( argc, ( const char * const * ) argv );

   hb_xexit();

   return iResult;
}

#if defined( HB_OS_WIN_CE ) && ! defined( __CEGCC__ )
   #include "hbwmain.c"
#endif
